### PR TITLE
Added Redis cache failover

### PR DESCRIPTION
**Pull request by vincenzodnp**
As described in issue #69 , this PR adds the cache failover if Redis endpoint is not available anymore and the Redis module is enabled.

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -100,14 +103,14 @@ services:
...
-    environment:
-      << : *default-environment
+#  solr:
+#    image: amazeeio/solr:6.6-drupal
+#    labels:
+#      lagoon.type: solr
+#    ports:
+#      - "8983" # exposes the port 8983 with a random local port, find it with `docker-compose port solr 8983`
+#    environment:
+#      << : *default-environment
```
Comment (seanhamlin) [inline docker-compose.yml]: Why is this commented out? Seems unrelated to Redis

File: web/sites/default/settings.php [hunk 2]
```diff
@@ -53,21 +53,34 @@
...
-  $settings['cache_prefix']['default'] = getenv('LAGOON_PROJECT') . '_' . getenv('LAGOON_GIT_SAFE_BRANCH');
-
-  # Do not set the cache during installations of Drupal.
-  if (!drupal_installation_attempted() && extension_loaded('redis')) {
     $settings['cache']['default'] = 'cache.backend.redis';
 
     // Include the default example.services.yml from the module, which will
     // replace all supported backend services (that currently includes the cache tags
-    // checksum service and the lock backends, check the file for the current list).
+    // checksum service and the lock backends, check the file for the current list)
```
Comment (seanhamlin) [inline web/sites/default/settings.php]: unrelated change? Drupal coding standards mandate the use of fullstops.

File: web/sites/default/settings.php [hunk 3]
```diff
@@ -107,6 +120,10 @@
       ],
     ];
   }
+  catch (\Exception $error) {
+    $settings['container_yamls'][] = 'sites/default/redis-unavailable.services.yml';
```
Comment (seanhamlin) [inline web/sites/default/settings.php:128]: I don't see this file in this PR https://github.com/govCMS/govcms8lagoon/blob/develop/.docker/images/govcms8/settings/redis-unavailable.services.yml

**Review by seanhamlin — CHANGES_REQUESTED** [review]

File: .lagoon.env [hunk 4]
```diff
@@ -0,0 +1 @@
+ENABLE_REDIS="true"
```
Comment (Schnitzel) [inline .lagoon.env]: let's not create a new environment variable, but instead remove the `getenv('ENABLE_REDIS')` check in the settings.php

**Review by Schnitzel — CHANGES_REQUESTED** [review]

File: docker-compose.yml (hunk 1 above)
Comment (vincenzodnp) [reply docker-compose.yml]: It was a leftover of my local setup. fixed

File: web/sites/default/settings.php (hunk 2 above)
Comment (vincenzodnp) [reply web/sites/default/settings.php]: Fixed.

File: web/sites/default/settings.php (hunk 3 above)
Comment (vincenzodnp) [reply web/sites/default/settings.php:128]: File added. Fixed

File: .lagoon.env (hunk 4 above)
Comment (vincenzodnp) [reply .lagoon.env]: Check removed. Fixed

**Review by vincenzodnp — COMMENTED** [review]
All requested reviews were fixed

**Review by Schnitzel — APPROVED** [review]

**Review by seanhamlin — CHANGES_REQUESTED** [review]
So, from my testing, I can confirm

* If Redis is down (e.g. scaled to 0, or the pod is deleted) then the sites will just spin and wait for the pod to come back online. The pages will work eventually.

I believe this behaviour is due to https://github.com/phpredis/phpredis#connect-open in which there is unlimited time spent waiting for the pod during the statement

```php
$redis->connect($redis_host, $redis_port);
```

I thing I think would be potentially useful is to amend it to

```php
$redis->connect($redis_host, $redis_port, 1);
```

This will force a 1 second timeout, and will allow pages to load with a null cache backend. I believe this is the original intent of this PR.

**Comment by vincenzodnp** [comment]
> This will force a 1 second timeout, and will allow pages to load with a null cache backend. I believe this is the original intent of this PR.

Currently, the PR-70 has timeout set, but no difference. 
Without Redis available, pages load but there is `60s` time somewhere set.
By checking the time of loading via Chrome developer's tool, page loads after 60 seconds.
Set the timeout doesn't help, but it's good to have it set

**Comment by vincenzodnp** [comment]
@seanhamlin I've re-deployed everything from scratch, and now it works as expected.
No more 60s timeout.

**Review by vincenzodnp — COMMENTED** [review]
Added the timeout as per your reques

**Comment by vincenzodnp** [comment]
@seanhamlin can you please review it?

**Comment by vincenzodnp** [comment]
@Schnitzel can we merge it?

File: web/sites/default/settings.php [hunk 5]
```diff
@@ -53,16 +53,33 @@
...
-  $settings['redis.connection']['interface'] = 'PhpRedis';
-  $settings['redis.connection']['host'] = getenv('REDIS_HOST') ?: 'redis';
-  $settings['redis.connection']['port'] = 6379;
-
-  $settings['cache_prefix']['default'] = getenv('LAGOON_PROJECT') . '_' . getenv('LAGOON_GIT_SAFE_BRANCH');
+// Redis configuration.
+if (getenv('LAGOON')) {
+  $redis = new \Redis();
+  $redis_host = getenv('REDIS_HOST') ?: 'redis';
+  $redis_port = getenv('REDIS_SERVICE_PORT') ?: 6379;
```
Comment (seanhamlin) [inline web/sites/default/settings.php:60]:
```suggestion
  $redis_port = getenv('REDIS_PORT') ?: 6379;
```

I like this for the env var a lot better.

**Review by seanhamlin — CHANGES_REQUESTED** [review]
As it turns out, we have a production failure of ElastiCache last Friday on a large cluster, and this in turn lead to connection pileups in ElastiCache. This took out every single site connecting to that ElastiCache cluster, until the ElastiCache cluster was failed over.

This cluster was already running a variant of this code.

So my thoughts in a nutshell:

* Using Redis in a pod is likely a bit safer for sites that are not under a lot of traffic (so small to medium)
* Using ElastiCache, you should not overload the amount of sites connecting to it, ideally have a single cluster per site (if possible), or use ElastiCache sharding.

I still have no proof that if Redis is not running, the site will continue to load, even with a null cache backend. This code does have a timeout on the connection, so this code does have the potential of working better.

In general, I have the one nitpick above, and let's just get this PR in, this code is at least a good headstart for people using Redis.

File: web/sites/default/settings.php [hunk 6]
```diff
@@ -53,16 +53,33 @@
...
-  $settings['redis.connection']['interface'] = 'PhpRedis';
-  $settings['redis.connection']['host'] = getenv('REDIS_HOST') ?: 'redis';
-  $settings['redis.connection']['port'] = 6379;
-
-  $settings['cache_prefix']['default'] = getenv('LAGOON_PROJECT') . '_' . getenv('LAGOON_GIT_SAFE_BRANCH');
+// Redis configuration.
+if (getenv('LAGOON')) {
+  $redis = new \Redis();
+  $redis_host = getenv('REDIS_HOST') ?: 'redis';
+  $redis_port = getenv('REDIS_PORT') ?: 6379;
```
Comment (Schnitzel) [inline web/sites/default/settings.php]: why this change? this will not work:
```
[drupal-example]master@cli-drupal:/app$ env | grep REDIS_PORT
REDIS_PORT=tcp://172.30.22.201:6379
```

**Review by Schnitzel — CHANGES_REQUESTED** [review]

File: web/sites/default/settings.php (hunk 6 above)
Comment (vincenzodnp) [reply web/sites/default/settings.php]: Reverted. Maybe it makes sense to change the name of the `REDIS_PORT` variable since it might lead to misunderstanding.

Comment (Schnitzel) [reply web/sites/default/settings.php]: mhh what to you mean with change the name of the variable? the variable is auto generated by kubernetes based on the name of the service, so we cannot really change the name.

Comment (vincenzodnp) [reply web/sites/default/settings.php]: Ok I see. I thought it was something Lagoon related.
I expect a variable called `REDIS_PORT` contains only a port number and not a full address, protocol included.
Can we merge the PR now?

Comment (Schnitzel) [reply web/sites/default/settings.php]: @vincenzodnp like sean said we don't have proof yet that this code actually works, I'm working on proofing this now.

**Comment by Schnitzel** [comment]
I did some tests to proof that the site actually still works even with redis not available, here the things I tested:

- delete the redis pod and let a new one start
- trigger a deployment of the redis deploymentconfig
- remove the redis service
- change the redis service to listen to an unknown port `6380` instead `6379`
- change the redis service to point to a port on the redis pod that is wrong 
- change the redis service to point to nginx instead of redis (will allow TCP connection, but the redis ping will fail).

during all of these tests I ran 
```
while true; do curl -s -I -w "Connect time: %{time_connect} Time to first byte: %{time_starttransfer} Total time: %{time_total} \n" https://nginx-drupal-example-pr-70.us.amazee.io/ | grep -E "Connect|HTTP"; sleep 1; done
```

which always gave me an output like this:

```
HTTP/1.1 200 OK
Connect time: 0.023469 Time to first byte: 0.171563 Total time: 0.171588
HTTP/1.1 200 OK
Connect time: 0.023796 Time to first byte: 0.161081 Total time: 0.161132
HTTP/1.1 200 OK
Connect time: 0.026849 Time to first byte: 0.147100 Total time: 0.147136
HTTP/1.1 200 OK
Connect time: 0.022869 Time to first byte: 0.150323 Total time: 0.150351
HTTP/1.1 200 OK
Connect time: 0.028385 Time to first byte: 0.153584 Total time: 0.153610
HTTP/1.1 200 OK
Connect time: 0.022464 Time to first byte: 0.157842 Total time: 0.157872
HTTP/1.1 200 OK
Connect time: 0.030782 Time to first byte: 0.147887 Total time: 0.147908
HTTP/1.1 200 OK
Connect time: 0.024848 Time to first byte: 0.153025 Total time: 0.153043
HTTP/1.1 200 OK
Connect time: 0.031318 Time to first byte: 0.156355 Total time: 0.156376 <--- test starts
HTTP/1.1 200 OK
Connect time: 0.022680 Time to first byte: 1.694557 Total time: 1.694613 <--- slower request
HTTP/1.1 200 OK
Connect time: 0.023541 Time to first byte: 1.029709 Total time: 1.029741
HTTP/1.1 200 OK
Connect time: 0.035552 Time to first byte: 1.746193 Total time: 1.746215
HTTP/1.1 200 OK
Connect time: 0.033998 Time to first byte: 0.941185 Total time: 0.941225
HTTP/1.1 200 OK
Connect time: 0.020820 Time to first byte: 1.613425 Total time: 1.613452
HTTP/1.1 200 OK
Connect time: 0.023161 Time to first byte: 1.041958 Total time: 1.041995
HTTP/1.1 200 OK
Connect time: 0.022668 Time to first byte: 1.649999 Total time: 1.650046
HTTP/1.1 200 OK
Connect time: 0.023615 Time to first byte: 1.042227 Total time: 1.042254
HTTP/1.1 200 OK
Connect time: 0.029577 Time to first byte: 1.606918 Total time: 1.606958
HTTP/1.1 200 OK
Connect time: 0.036274 Time to first byte: 1.041331 Total time: 1.041355
HTTP/1.1 200 OK
Connect time: 0.104599 Time to first byte: 1.883088 Total time: 1.883111
HTTP/1.1 200 OK
Connect time: 0.029954 Time to first byte: 0.861338 Total time: 0.861364
HTTP/1.1 200 OK
Connect time: 0.029168 Time to first byte: 0.181294 Total time: 0.181323 <--- test ends, everything normal
HTTP/1.1 200 OK
Connect time: 0.028094 Time to first byte: 0.188738 Total time: 0.188804
HTTP/1.1 200 OK
Connect time: 0.028617 Time to first byte: 0.141155 Total time: 0.141182
HTTP/1.1 200 OK
Connect time: 0.023308 Time to first byte: 0.157814 Total time: 0.157840
HTTP/1.1 200 OK
Connect time: 0.024179 Time to first byte: 0.167849 Total time: 0.167875
HTTP/1.1 200 OK
Connect time: 0.024269 Time to first byte: 0.176354 Total time: 0.176382
HTTP/1.1 200 OK
Connect time: 0.024329 Time to first byte: 0.151168 Total time: 0.151191
HTTP/1.1 200 OK
Connect time: 0.030315 Time to first byte: 0.159573 Total time: 0.159602
HTTP/1.1 200 OK
```

Plus I wanted to check that the Drupal is actually not slower having this additional redis test in there, so I did two tests:

1. run 1000 requests (1 concurrent) against the nginx pod with the redis uptime check removed (basically the code before this PR)
2. run 1000 requests (1 concurrent) against the nginx pod with the redis uptime check introduced in this PR,

the results are good:

Nr 1 (no redis check):
```
Time per request:       47.997 [ms] (mean, across all concurrent requests)
Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        5    7   4.2      6      76
Processing:    31   41   9.2     38     131
Waiting:       31   41   9.2     38     131
Total:         36   48  10.4     45     166
```

Nr 2 (with redis check):
```
Time per request:       48.375 [ms] (mean, across all concurrent requests)
Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        5    7   3.4      6      56
Processing:    31   41  10.6     39     241
Waiting:       31   41  10.6     39     241
Total:         38   48  11.2     46     248
```

this looks like being 0.4ms slower with the redis check, but this could also just be something else.

From my side therefore good to merge, what do you think @seanhamlin

**Review by Schnitzel — APPROVED** [review]

**Comment by seanhamlin** [comment]
Wow, great analysis. I think this at least proves the 1 second timeout is working, as the time hovers around 1 second with redis down. Let's get this in.
